src/js/components/ServiceForm.js

```diff
@@ -0,0 +1,230 @@
+import {Tooltip} from 'reactjs-components';
+import React from 'react';
+
+import FormUtil from '../utils/FormUtil';
+import GeminiUtil from '../utils/GeminiUtil';
+import SchemaFormUtil from '../utils/SchemaFormUtil';
+import SchemaUtil from '../utils/SchemaUtil';
+import TabForm from './TabForm';
+import Util from '../utils/Util';
+
+const METHODS_TO_BIND = [
+  'validateForm', 'getTriggerTabFormSubmit', 'handleFormChange',
+  'handleExternalSubmit', 'getRemoveVariableButton', 'getAddNewRowButton'
+];
+
+class ServiceForm extends React.Component {
+  constructor() {
+    super();
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+
+    this.triggerSubmit = function () {};
+    this.isValidated = true;
+  }
+
+  componentWillMount() {
+    this.multipleDefinition = this.getNewDefinition();
+    this.props.getTriggerSubmit(this.handleExternalSubmit);
+  }
+
+  componentWillUnmount() {
+    // Unschedule all validation if component unmounts.
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+  }
+
+  componentDidUpdate() {
+    // Timeout necessary due to modal content height updates on did mount
+    setTimeout(() => {
+      GeminiUtil.updateWithRef(this.refs.geminiForms);
+    });
+  }
+
+  handleFormChange(formData, eventObj) {
+    let isBlur = eventObj.eventType === 'blur';
+    let isChange = eventObj.eventType === 'change';
+    if (!isBlur && !isChange) {
+      return;
+    }
+
+    if (isBlur) {
+      this.validateForm();
+      this.props.onChange(this.getModel());
+      return;
+    }
+
+    // The cleartimeout is there to debounce the validation. And to make
+    // sure it is only run once.
+    if (this.timer) {
+      clearTimeout(this.timer);
+    }
+    this.timer = setTimeout(() => {
+      this.validateForm();
+      this.props.onChange(this.getModel());
+    });
+  }
+
+  handleExternalSubmit() {
+    this.validateForm();
+    return this.getModel();
+  }
+
+  handleRemoveRow(definition, prop, id) {
+    FormUtil.removePropID(definition, prop, id);
+    this.forceUpdate();
+  }
+
+  handleAddRow(prop, definition, newDefinition) {
+    let propID = Util.uniqueID(prop);
+    newDefinition = FormUtil.getMultipleFieldDefinition(
+      prop,
+      propID,
+      newDefinition
+    );
+    newDefinition.push(
+      this.getRemoveVariableButton(definition, prop, propID)
+    );
+    definition.splice(definition.length - 1, 0, newDefinition);
+    this.forceUpdate();
+  }
+
+  transformPorts(definition, startValues) {
```

Comment:
This is tight coupled with one definition, is it possible to generalize this, I would like to have this for all array fields.